File: text.go
package bdf

import "math"

// MARK kinds (docs/spec.md §7.8).
const (
	MarkParagraph byte = 0
	MarkLine      byte = 1
	MarkCell      byte = 2
	MarkBox       byte = 3
	MarkAltText   byte = 4
	MarkWrap      byte = 5
	MarkHeading   byte = 6  // payload: level "1"-"6"
	MarkList      byte = 7  // opens a list, closed by MarkEnd
	MarkListItem  byte = 8  // starts an item of the innermost list
	MarkTable     byte = 9  // opens a table, closed by MarkEnd
	MarkFigure    byte = 10 // payload: alternative text; opens a figure, closed by MarkEnd
	MarkEnd       byte = 11 // closes the innermost list, table or figure
	MarkLang      byte = 12 // payload: BCP 47 language of the runs that follow ("" = document default)
)

// Separators between consecutive text runs (docs/spec.md §7.9).
const (
	SepNone  byte = 0
	SepSpace byte = 1
	SepBreak byte = 2
)

// TextRun is a text drawing instruction (or ALT_TEXT mark) located in the
// coordinate space of the object being walked.
type TextRun struct {
	Text    string
	X, Y    float32 // anchor after applying the transform
	Advance float32
	Size    float32
	Font    *Font
	Align   byte
	Matrix  [6]float32
	Sep     byte // separator from the previous run, from MARKs or position heuristics
	Ordinal int  // position in the extraction order of the top-level object
	AltText bool
}

type matrix [6]float32

func mul(m, n matrix) matrix {
	return matrix{
		m[0]*n[0] + m[2]*n[1],
		m[1]*n[0] + m[3]*n[1],
		m[0]*n[2] + m[2]*n[3],
		m[1]*n[2] + m[3]*n[3],
		m[0]*n[4] + m[2]*n[5] + m[4],
		m[1]*n[4] + m[3]*n[5] + m[5],
	}
}

type textState struct {
	m     matrix
	font  *Font
	size  float32
	align byte
}

type textExtractor struct {
	runs    []TextRun
	resolve func(Hash) *ObjectPart
	pending byte // separator requested by a MARK for the next run
	hasMark bool
	alt     *string // ALT_TEXT waiting for the drawing op it describes
}

func (t *textExtractor) walk(o *ObjectPart, m matrix) error {
	st := textState{m: m, size: 10}
	var stack []textState
	return o.Walk(func(in Instr) {
		switch in.Op {
		case OpSave:
			stack = append(stack, st)
		case OpRestore:
			if n := len(stack); n > 0 {
				st = stack[n-1]
				stack = stack[:n-1]
			}
		case OpTransform:
			st.m = mul(st.m, matrix{f(in, 0), f(in, 1), f(in, 2), f(in, 3), f(in, 4), f(in, 5)})
		case OpTranslate:
			st.m = mul(st.m, matrix{1, 0, 0, 1, f(in, 0), f(in, 1)})
		case OpScale:
			st.m = mul(st.m, matrix{f(in, 0), 0, 0, f(in, 1), 0, 0})
		case OpFont:
			if i := int(in.Args[0].(uint64)); i < len(o.Fonts) {
				st.font = &o.Fonts[i]
			}
			st.size = f(in, 1)
		case OpTextStyle:
			st.align = byte(in.Args[0].(uint64))
		case OpMark:
			kind := byte(in.Args[0].(uint64))
			payload := in.Args[1].(string)
			switch kind {
			case MarkLine:
				t.flushAlt(st)
				t.mark(SepSpace)
			case MarkParagraph, MarkCell, MarkBox, MarkHeading, MarkList, MarkListItem, MarkTable, MarkFigure, MarkEnd:
				t.flushAlt(st)
				t.mark(SepBreak)
			case MarkWrap:
				t.flushAlt(st)
				t.mark(SepNone)
			case MarkAltText:
				t.flushAlt(st)
				p := payload
				t.alt = &p
			}
			// MarkLang and unknown kinds leave the separator to the next MARK
			// or to the position guess.
		case OpFillText, OpStrokeText:
			if t.alt != nil {
				// The drawing op right after ALT_TEXT renders that text.
				t.emit(*t.alt, f(in, 1), f(in, 2), f(in, 3), st, true)
				t.alt = nil
				return
			}
			t.emit(in.Args[0].(string), f(in, 1), f(in, 2), f(in, 3), st, false)
		case OpFillPathAt:
			if t.alt != nil {
				t.emit(*t.alt, f(in, 2), f(in, 3), 0, st, true)
				t.alt = nil
			}
		case OpFillPathRun:
			if t.alt != nil {
				if gl := in.Args[1].([]Glyph); len(gl) > 0 {
					t.emit(*t.alt, gl[0].X, gl[0].Y, 0, st, true)
				} else {
					t.emit(*t.alt, 0, 0, 0, st, true)
				}
				t.alt = nil
			}
		case OpUse, OpUseAt:
			if t.alt != nil {
				x, y := float32(0), float32(0)
				if in.Op == OpUseAt {
					x, y = f(in, 1), f(in, 2)
				}
				t.emit(*t.alt, x, y, 0, st, true)
				t.alt = nil
				return
			}
			child := t.resolve(o.Objects[int(in.Args[0].(uint64))])
			if child == nil {
				return
			}
			cm := st.m
			if in.Op == OpUseAt {
				cm = mul(cm, matrix{1, 0, 0, 1, f(in, 1), f(in, 2)})
			}
			_ = t.walk(child, cm)
		}
	})
}

func f(in Instr, i int) float32 { return in.Args[i].(float32) }

// flushAlt emits an ALT_TEXT that no drawing op consumed, without a position.
func (t *textExtractor) flushAlt(st textState) {
	if t.alt != nil {
		t.emit(*t.alt, 0, 0, 0, st, true)
		t.alt = nil
	}
}

func (t *textExtractor) mark(sep byte) {
	if !t.hasMark || sep > t.pending {
		t.pending = sep
	}
	t.hasMark = true
}

func (t *textExtractor) emit(text string, x, y, advance float32, st textState, alt bool) {
	m := st.m
	run := TextRun{
		Text: text, Advance: advance, Size: st.size, Font: st.font, Align: st.align, Matrix: m,
		X: m[0]*x + m[2]*y + m[4], Y: m[1]*x + m[3]*y + m[5], Ordinal: len(t.runs), AltText: alt,
	}
	if t.hasMark {
		run.Sep = t.pending
	} else if len(t.runs) > 0 {
		run.Sep = guessSep(t.runs[len(t.runs)-1], run)
	}
	t.pending, t.hasMark = SepNone, false
	t.runs = append(t.runs, run)
}

// guessSep estimates the separator between runs that carry no MARK information.
func guessSep(prev, cur TextRun) byte {
	size := prev.Size
	if size <= 0 {
		size = 10
	}
	if math.Abs(float64(cur.Y-prev.Y)) > float64(size)*0.5 {
		return SepSpace
	}
	end := prev.X + prev.Advance
	if prev.Advance == 0 || cur.X-end > float64To32(0.2)*size {
		return SepSpace
	}
	return SepNone
}

func float64To32(v float64) float32 { return float32(v) }

// ExtractText returns the text runs of an object and the objects it USEs, in
// walk order. resolve must return decoded child objects (nil skips the child).
func ExtractText(o *ObjectPart, resolve func(Hash) *ObjectPart) ([]TextRun, error) {
	t := &textExtractor{resolve: resolve}
	err := t.walk(o, matrix{1, 0, 0, 1, 0, 0})
	t.flushAlt(textState{m: matrix{1, 0, 0, 1, 0, 0}, size: 10})
	return t.runs, err
}

File: packages/core/src/text.ts
import { NoopSink, walk } from "./object.js";
import type { ObjectPart, Font, Hash } from "./types.js";

/** MARK kinds (docs/spec.md §7.8). */
export const Mark = { PARAGRAPH: 0, LINE: 1, CELL: 2, BOX: 3, ALT_TEXT: 4, WRAP: 5 } as const;

/** Separators between consecutive runs (docs/spec.md §7.9). */
export const Sep = { NONE: 0, SPACE: 1, BREAK: 2 } as const;

/** A text run in the coordinate space passed to extractText. */
export interface TextRun {
  text: string;
  /** Anchor position after applying the transform. */
  x: number;
  y: number;
  /** Advance in the object's units (0 when unknown). */
  advance: number;
  size: number;
  font: Font | undefined;
  align: number;
  /** Full 2D transform at the time of drawing (a b c d e f). */
  matrix: [number, number, number, number, number, number];
  /** Separator from the previous run (Sep.*). */
  sep: number;
  /** Position in the extraction order of the top-level object; matches the text index. */
  ordinal: number;
  /** True for ALT_TEXT marks (nothing was drawn with fillText). */
  altText: boolean;
}

type Matrix = [number, number, number, number, number, number];

function mul(m: Matrix, n: Matrix): Matrix {
  return [
    m[0] * n[0] + m[2] * n[1],
    m[1] * n[0] + m[3] * n[1],
    m[0] * n[2] + m[2] * n[3],
    m[1] * n[2] + m[3] * n[3],
    m[0] * n[4] + m[2] * n[5] + m[4],
    m[1] * n[4] + m[3] * n[5] + m[5],
  ];
}

interface State { m: Matrix; font: Font | undefined; size: number; align: number }

/** Shared run list and MARK state across nested objects. */
class Extraction {
  runs: TextRun[] = [];
  pending = Sep.NONE as number;
  hasMark = false;
  /** ALT_TEXT waiting for the drawing op it describes. */
  alt: string | undefined;
  constructor(readonly resolve: (h: Hash) => ObjectPart | undefined) {}

  mark(sep: number) {
    if (!this.hasMark || sep > this.pending) this.pending = sep;
    this.hasMark = true;
  }

  /** Emit an ALT_TEXT that no drawing op consumed, without a position. */
  flushAlt(st: State) {
    if (this.alt !== undefined) {
      const t = this.alt;
      this.alt = undefined;
      this.emit(t, 0, 0, 0, st, true);
    }
  }

  emit(text: string, x: number, y: number, advance: number, st: State, altText: boolean) {
    const m = st.m;
    const run: TextRun = {
      text, advance, size: st.size, font: st.font, align: st.align, matrix: m, altText,
      x: m[0] * x + m[2] * y + m[4], y: m[1] * x + m[3] * y + m[5],
      sep: Sep.NONE, ordinal: this.runs.length,
    };
    if (this.hasMark) run.sep = this.pending;
    else if (this.runs.length) run.sep = guessSep(this.runs[this.runs.length - 1], run);
    this.pending = Sep.NONE;
    this.hasMark = false;
    this.runs.push(run);
  }
}

/** Estimate the separator between runs that carry no MARK information. */
export function guessSep(prev: TextRun, cur: TextRun): number {
  const size = prev.size > 0 ? prev.size : 10;
  if (Math.abs(cur.y - prev.y) > size * 0.5) return Sep.SPACE;
  const end = prev.x + prev.advance;
  if (prev.advance === 0 || cur.x - end > Math.fround(0.2) * size) return Sep.SPACE;
  return Sep.NONE;
}

class TextSink extends NoopSink {
  private stack: State[] = [];
  private st: State;

  constructor(private ex: Extraction, private obj: ObjectPart, m: Matrix) {
    super();
    this.st = { m, font: undefined, size: 10, align: 0 };
  }
  override save() { this.stack.push({ ...this.st }); }
  override restore() { this.st = this.stack.pop() ?? this.st; }
  override transform(a: number, b: number, c: number, d: number, e: number, f: number) { this.st.m = mul(this.st.m, [a, b, c, d, e, f]); }
  override translate(x: number, y: number) { this.transform(1, 0, 0, 1, x, y); }
  override scale(x: number, y: number) { this.transform(x, 0, 0, y, 0, 0); }
  override font(font: number, size: number) { this.st.font = this.obj.fonts[font]; this.st.size = size; }
  override textStyle(align: number) { this.st.align = align; }
  override mark(kind: number, payload: string) {
    switch (kind) {
      case Mark.LINE: this.ex.flushAlt(this.st); this.ex.mark(Sep.SPACE); break;
      case Mark.PARAGRAPH: case Mark.CELL: case Mark.BOX: this.ex.flushAlt(this.st); this.ex.mark(Sep.BREAK); break;
      case Mark.ALT_TEXT: this.ex.flushAlt(this.st); this.ex.alt = payload; break;
      case Mark.WRAP: this.ex.flushAlt(this.st); this.ex.mark(Sep.NONE); break;
    }
  }
  /** The drawing op right after ALT_TEXT renders that text; returns true when consumed. */
  private takeAlt(x: number, y: number, advance: number): boolean {
    if (this.ex.alt === undefined) return false;
    const t = this.ex.alt;
    this.ex.alt = undefined;
    this.ex.emit(t, x, y, advance, this.st, true);
    return true;
  }
  override fillText(text: string, x: number, y: number, advance: number) { if (!this.takeAlt(x, y, advance)) this.ex.emit(text, x, y, advance, this.st, false); }
  override strokeText(text: string, x: number, y: number, advance: number) { if (!this.takeAlt(x, y, advance)) this.ex.emit(text, x, y, advance, this.st, false); }
  override fillPathAt(_path: number, _rule: number, x: number, y: number) { this.takeAlt(x, y, 0); }
  override fillPathRun(_rule: number, glyphs: { x: number; y: number }[]) { this.takeAlt(glyphs[0]?.x ?? 0, glyphs[0]?.y ?? 0, 0); }
  override use(obj: number) { this.useAt(obj, 0, 0); }
  override useAt(obj: number, x: number, y: number) {
    if (this.takeAlt(x, y, 0)) return;
    const child = this.ex.resolve(this.obj.objects[obj]);
    if (!child) return;
    walk(child, new TextSink(this.ex, child, mul(this.st.m, [1, 0, 0, 1, x, y])));
  }
}

/**
 * Extract text runs from an object (and the objects it USEs) in the space
 * defined by matrix. resolve must return already-loaded child objects.
 * The order and ordinals match the Go extractor and the text index part.
 */
export function extractText(obj: ObjectPart, resolve: (h: Hash) => ObjectPart | undefined, matrix: Matrix = [1, 0, 0, 1, 0, 0]): TextRun[] {
  const ex = new Extraction(resolve);
  walk(obj, new TextSink(ex, obj, matrix));
  ex.flushAlt({ m: matrix, font: undefined, size: 10, align: 0 });
  return ex.runs;
}
